lib/task/import/csvAuthorityRecordRelationImportTask.class.php

```diff
@@ -218,7 +232,96 @@ private function relationshipExists($sourceActorId, $targetActorId, $relationTyp
       ':type_id' => $relationTypeId
     ];
 
-    return QubitPdo::fetchOne($sql, $params) !== false
-      || QubitPdo::fetchOne($sql, $paramsVariant) !== false;
+    if ($relationId = QubitPdo::fetchColumn($sql, $params))
+    {
+      return $relationId;
+    }
+
+    $relationId = QubitPdo::fetchColumn($sql, $paramsVariant);
+
+    return !empty($relationId) ? $relationId : false;
+  }
+
+  /**
+   * Add relation between two actors
+   *
+   * @param int $sourceActorId  ID of an actor
+   * @param int $targetActorId  ID of an actor
+   * @param int $relationTypeId  ID of relation type
+   */
+  private function addRelation($sourceActorId, $targetActorId, $relationTypeId)
+  {
+    $relation = new QubitRelation;
+    $this->setRelationFields($relation, $sourceActorId, $targetActorId, $relationTypeId);
+    $relation->save();
+
+    $this->noteActorIds($sourceActorId, $targetActorId);
+
+    // Add keymap entry
+    if (!empty($this->import->columnValue('legacyId')))
+    {
+      $this->import->createKeymapEntry($this->import->getStatus('sourceName'), $this->import->columnValue('legacyId'), $relation);
+    }
+  }
+
+  /**
+   * Update relation using CSV row data
+   *
+   * @param int $relationId  ID of relation to update
+   * @param int $sourceActorId  ID of an actor
+   * @param int $targetActorId  ID of an actor
+   * @param int $relationTypeId  ID of relation type
+   */
+  private function updateRelation($relationId, $sourceActorId, $targetActorId, $relationTypeId)
+  {
+    $relation = QubitRelation::getById($relationId);
+    $this->setRelationFields($relation, $sourceActorId, $targetActorId, $relationTypeId);
+    $relation->save();
+
+    $this->noteActorIds($sourceActorId, $targetActorId);
+  }
+
+  /**
+   * Set relation properties using CSV row data
+   *
+   * @param QubitRelation $relation  Relation to update
+   * @param int $sourceActorId  ID of an actor
+   * @param int $targetActorId  ID of an actor
+   * @param int $relationTypeId  ID of relation type
+   */
+  private function setRelationFields(&$relation, $sourceActorId, $targetActorId, $relationTypeId)
+  {
+    $relation->objectId  = $sourceActorId;
+    $relation->subjectId = $targetActorId;
+    $relation->typeId    = $relationTypeId;
+
+    // Set relationship properties from column values
+    foreach (['date', 'startDate', 'endDate', 'description'] as $property)
+    {
+      if (!empty($this->import->columnValue($property)))
+      {
+        $relation->$property = $this->import->columnValue($property);
+      }
+    }
+  }
+
+  /**
+   * Note actor IDs whose relationship data has possibly changed
+   *
+   * @param int $sourceActorId  ID of an actor
+   * @param int $targetActorId  ID of an actor
+   */
+  private function noteActorIds($sourceActorId, $targetActorId)
```

Comment:
This would be more concise if the method took an array of actor ids as an argument (e.g. `noteActorIds(array $actorIds)`) then iterated through the array. The logic is the same for source or target ids, so you could just pass them both as array members.

Nitpick: When I saw "noteActorIds" I thought this was related to the 'note' table. 😆 Maybe a name like `addUpdatedActorIds()` or `addUpdatedIds()` would avoid potential confusion?